File: 1460-number-of-substrings-containing-all-three-characters/1460-number-of-substrings-containing-all-three-characters.java
class Solution {
  // Similar to 3. Longest SubString Without Repeating Characters
  public int numberOfSubstrings(String s) {
    int ans = 0;
    int[] lastSeen = new int[3]; // lastSeen[c] := index of last c appeared.
    Arrays.fill(lastSeen, -1);

    for (int i = 0; i < s.length(); ++i) {
      lastSeen[s.charAt(i) - 'a'] = i;
      // s[0..i], s[1..i], s[Math.min(lastSeen)..i] are satisfied strings.
      ans += 1 + Arrays.stream(lastSeen).min().getAsInt();
    }

    return ans;
  }
}

File: 1460-number-of-substrings-containing-all-three-characters/1460-number-of-substrings-containing-all-three-characters.cpp
class Solution {
 public:
  // Similar to 3. Longest Substring Without Repeating Characters
  int numberOfSubstrings(string s) {
    int ans = 0;
    vector<int> lastSeen(3, -1);  // lastSeen[c] := index of last c appeared.

    for (int i = 0; i < s.length(); ++i) {
      lastSeen[s[i] - 'a'] = i;
      // s[0..i], s[1..i], s[min(lastSeen)..i] are satisfied strings.
      ans += 1 + *min_element(lastSeen.begin(), lastSeen.end());
    }

    return ans;
  }
};
